File: custom_msg_srv/src/test_custom_interfaces/src/publisher_member_function.cpp
#include <functional>
#include <memory>
#include <chrono>
#include "rclcpp/rclcpp.hpp"
#include "custom_msg_srv/msg/num.hpp"

using namespace std::chrono_literals;

class MinimalPublisher : public rclcpp::Node
{
  public:

    MinimalPublisher() : Node("num_publisher"){

      count_ = 0;

      publisher_ = this->create_publisher<custom_msg_srv::msg::Num>("my_num",10);

      timer_ = this->create_wall_timer(1s, std::bind(&MinimalPublisher::timer_callback, this));
      RCLCPP_INFO(this->get_logger(), "Publishing at 1Hz");
    }

  private:

    size_t count_;
    rclcpp::Publisher<custom_msg_srv::msg::Num>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;


    void timer_callback(){

      auto message = custom_msg_srv::msg::Num();
      message.num = count_;

      RCLCPP_INFO(this->get_logger(), "Publishing: %ld", message.num);
      this->publisher_->publish(message);

      count_++;
    }
};


int main(int argc, char* argv[]){

  rclcpp::init(argc, argv);


  auto node = std::make_shared<MinimalPublisher>();
  rclcpp::spin(node);

  rclcpp::shutdown();

  return 0;
}

File: custom_msg_srv/src/test_custom_interfaces/test_custom_interfaces/publisher_member_function.py
#! /usr/bin/env python3

import rclpy  # import the ROS2 client lib for pyhton
from rclpy.node import Node  # import the Node class, used for creating nodes
from rclpy.executors import ExternalShutdownException
from custom_msg_srv.msg import Num


class MinimalPublisher(Node):

    def __init__(self):

        super().__init__("num_publisher")

        self._publisher = self.create_publisher(Num, "my_num", 10)

        self._timer_period = 1
        self._timer = self.create_timer(self._timer_period, self._timer_callback)

        self._counter = 0

    def _timer_callback(self):

        msg = Num()

        msg.num = self._counter

        self._publisher.publish(msg)

        self.get_logger().info(f'Publishing: "{msg}"')

        self._counter += 1


def main(args=None):

    try:

        rclpy.init(args=args)

        node = MinimalPublisher()

        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        print()
    except Exception as e:
        print(f"An error occurred: {e}")
    finally:
        if node is not None:
            # Destroy the node explicitly
            node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()  # Shutdown ROS 2 communication


if __name__ == "__main__":
    main()  # execute the main function if the script is run directly
